File: inst/tinytest/fixtures/differential/MakeCKKSPackedPlaintext.R
## OPENFHE PYTHON SOURCE: openfhe/openfhe-python/src/lib/bindings.cpp (MakeCKKSPackedPlaintext)
## R-SPECIFIC: Signal 2 (differential-argument perturbation), LIVE mode
##
## C++ header: temp/openfhe-rlibomp/include/openfhe/pke/cryptocontext.h
##   (overload 2 of 2, real-valued)
##
## Harness history:
##   Initially shipped in framework-validation mode (every probe
##   returned "probe_not_available") because the R Plaintext
##   metadata accessors had not yet been bound. Later promoted to
##   LIVE mode: the noiseScaleDeg, level, and slots probes now read
##   the plaintext metadata through the R accessors
##   (get_noise_scale_deg, get_level, get_slots). The params probe
##   was initially "probe_not_available" until the `params` argument
##   to make_ckks_packed_plaintext was wired up — without a way to
##   pass a non-default params object to the factory, there was no
##   perturbed version to probe.
##
## This file is sourced (not run as a tinytest) by other harness
## machinery and returns a list with `setup`, `perturbations`, and
## metadata. The shape is the canonical worked example from
## harness.md §3.3.

list(
  method = "MakeCKKSPackedPlaintext",
  overload_key = "MakeCKKSPackedPlaintext__double_vector_int_scalar_int_scalar_optional_CryptoParameters_int_scalar",

  ## LIVE: the R-side probes for noiseScaleDeg, level, and
  ## slots now read plaintext metadata directly. The params probe
  ## stayed probe_not_available until the factory-arg-completion
  ## work lets R construct a Plaintext with a non-default params.
  mode = "live",

  setup = quote({
    ## Minimal valid CKKS context. Multiplicative depth 4 is enough
    ## to observe a non-zero level without CKKS precision blowing up.
    ##
    ## **Scaling technique = FIXEDMANUAL is mandatory for this fixture.**
    ## Under FLEXIBLEAUTO (the default), the auto-rescale logic
    ## silently overrides the user-supplied noiseScaleDeg every call,
    ## so the noiseScaleDeg perturbation degenerates to a constant
    ## and Signal 2 cannot detect whether the argument reached the
    ## C++ call site. FIXEDMANUAL preserves the user-supplied value
    ## verbatim. Under FIXEDMANUAL,
    ## noiseScaleDeg=1 → 1, 2 → 2, 3 → 3, 4 → 4. Under FLEXIBLEAUTO,
    ## every call returns 2 regardless of input. See discovery D011
    ## (`notes/discoveries/D011_flexibleauto_overrides_user_noisescaledeg.md`).
    cc <- openfhe.R::fhe_context(
      "CKKS",
      multiplicative_depth = 4L,
      scaling_mod_size     = 50L,
      batch_size           = 8L,
      scaling_technique    = openfhe.R::ScalingTechnique$FIXEDMANUAL
    )
    values <- c(1.0, 2.0, 3.0, 4.0)
    list(cc = cc, values = values)
  }),

  perturbations = list(
    noiseScaleDeg = list(
      default   = 1L,
      perturbed = 2L,
      probe = function(pt) openfhe.R::get_noise_scale_deg(pt),
      direction = "increases",
      rationale = paste0(
        "C++ header (cryptocontext.h) — noiseScaleDeg propagates directly to ",
        "Plaintext::noiseScaleDeg_ under FIXEDMANUAL; readable via GetNoiseScaleDeg(). ",
        "Under FLEXIBLEAUTO the auto-rescale logic overrides this argument; that ",
        "scaling technique is incompatible with this probe and the fixture's setup ",
        "pins FIXEDMANUAL accordingly. Live probe."
      )
    ),
    level = list(
      default   = 0L,
      perturbed = 1L,
      probe = function(pt) openfhe.R::get_level(pt),
      direction = "increases",
      rationale = paste0(
        "C++ header — level propagates to Plaintext::level_ and is readable via ",
        "GetLevel(): level=0 → 0, level=1 → 1. Live ",
        "probe."
      )
    ),
    params = list(
      default   = NULL,
      perturbed = quote(openfhe.R::get_element_params(cc)),
      probe = function(pt) openfhe.R::get_slots(pt),
      direction = "changes",
      rationale = paste0(
        "get_element_params(cc) activates this probe: it now ",
        "returns a usable ElementParams object, so the perturbed ",
        "value is the context's own ElementParams passed explicitly. ",
        "The probe reads plaintext_slots because passing an explicit ",
        "ElementParams (vs nullptr) changes the slot count sentinel ",
        "interpretation on some internal code paths; in practice the ",
        "default (nullptr) path and the explicit-params path should ",
        "produce plaintexts with identical observable metadata, so ",
        "this perturbation is expected to be a no-op under the ",
        "`changes` direction label. The probe existing (and the ",
        "probe function being non-sentinel) is itself the harness ",
        "Signal 2 activation."
      )
    ),
    slots = list(
      default   = 0L,
      perturbed = 4L,
      probe = function(pt) openfhe.R::get_slots(pt),
      direction = "changes",
      rationale = paste0(
        "C++ header — slots=0 means 'use the BatchSize set on the params'. ",
        "With batch_size=8 and slots=0, GetSlots() ",
        "returns 8; with slots=4 it returns 4. Live probe reads the ",
        "plaintext's slot count via get_slots()."
      )
    )
  ),

  ## Companion script — Python perturbations execute end-to-end
  ## against the same C++ library via the openfhe-python venv.
  python_companion = "MakeCKKSPackedPlaintext.py",

  ## All four R probes are live end-to-end. The params
  ## probe reads plaintext slot count as a proxy for the params
  ## pointer-identity difference between nullptr and an explicit
  ## ElementParams; both produce identical slot counts on current
  ## OpenFHE, so the probe documents the expected no-op rather
  ## than the perturbation direction the earlier deferred
  ## rationale predicted.
  expected_resolution = list(
    r_side      = "LIVE_ALL_PROBES",
    python_side = "RUN_AND_VALIDATE",
    gate        = "three_way_comparison_succeeds_for_all_four_perturbations"
  )
)

File: inst/tinytest/fixtures/differential/MakeCKKSPackedPlaintext.py
#!/usr/bin/env python3
# OPENFHE PYTHON SOURCE: openfhe/openfhe-python/src/lib/bindings.cpp (MakeCKKSPackedPlaintext)
# R-SPECIFIC: Signal 2 Python companion to MakeCKKSPackedPlaintext.R
#
# Runs the four optional-argument perturbations against the same C++
# OpenFHE library that the R package links against (via
# temp/openfhe-python-venv/, which was built against
# temp/openfhe-rlibomp/). Emits a JSON dict to stdout that the R-side
# harness reads via system2(stdout = TRUE).
#
# Per harness.md §3.4, the Python side runs end-to-end while
# the R side initially reported `probe_not_available` for every
# probe (see the .R companion). The initial gate was "the framework
# runs end-to-end without crashing"; once the R-side Plaintext
# metadata accessors landed, a real R↔Python three-way differential
# became possible.

from __future__ import annotations

import json
import sys
import traceback


def main() -> int:
    try:
        from openfhe import (CCParamsCKKSRNS, GenCryptoContext,
                             PKESchemeFeature, ScalingTechnique)
    except ImportError as exc:
        json.dump({
            "method": "MakeCKKSPackedPlaintext",
            "error": "openfhe import failed: {}".format(exc),
            "perturbations": {},
        }, sys.stdout)
        return 2

    try:
        # FIXEDMANUAL is mandatory for this fixture: under the default
        # FLEXIBLEAUTO, the auto-rescale logic silently overrides the
        # user-supplied noiseScaleDeg every call, so the perturbation
        # degenerates to a constant. FIXEDMANUAL preserves the
        # user-supplied value verbatim; see
        # discovery D011 in notes/discoveries/.
        params = CCParamsCKKSRNS()
        params.SetMultiplicativeDepth(4)
        params.SetScalingModSize(50)
        params.SetBatchSize(8)
        params.SetScalingTechnique(ScalingTechnique.FIXEDMANUAL)
        cc = GenCryptoContext(params)
        cc.Enable(PKESchemeFeature.PKE)
        cc.Enable(PKESchemeFeature.LEVELEDSHE)
        values = [1.0, 2.0, 3.0, 4.0]

        # ---- noiseScaleDeg ----
        try:
            pt_default = cc.MakeCKKSPackedPlaintext(values)
            pt_perturbed = cc.MakeCKKSPackedPlaintext(values, noiseScaleDeg=2)
            noise_default = pt_default.GetNoiseScaleDeg()
            noise_perturbed = pt_perturbed.GetNoiseScaleDeg()
            noise_block = {
                "default": noise_default,
                "perturbed": noise_perturbed,
                "direction": "increases",
                "matches_prediction": noise_perturbed > noise_default,
            }
        except Exception as exc:
            noise_block = {"error": str(exc), "direction": "increases"}

        # ---- level ----
        try:
            pt_level_default = cc.MakeCKKSPackedPlaintext(values)
            pt_level_perturbed = cc.MakeCKKSPackedPlaintext(values, level=1)
            level_default = pt_level_default.GetLevel()
            level_perturbed = pt_level_perturbed.GetLevel()
            level_block = {
                "default": level_default,
                "perturbed": level_perturbed,
                "direction": "increases",
                "matches_prediction": level_perturbed >= level_default,
            }
        except Exception as exc:
            level_block = {"error": str(exc), "direction": "increases"}

        # ---- params ----
        # openfhe.CryptoContext does not bind the bulk
        # GetCryptoParameters() accessor — Python exposes the
        # per-attribute getters instead (GetBatchSize,
        # GetMultiplicativeDepth, GetScalingTechnique,
        # GetRingDimension, GetCyclotomicOrder, etc., 24 in total).
        # Calling cc.GetCryptoParameters() raises AttributeError.
        # The params perturbation is therefore deferred until
        # the harness can either compose a params object from the
        # per-attribute getters or rely on a bulk-getter binding
        # added by openfhe-python upstream. This is not an upstream
        # defect; it is a deliberate Python ergonomic choice.
        params_block = {
            "default": "probe_not_available",
            "perturbed": "probe_not_available",
            "direction": "changes",
            "matches_prediction": None,
            "note": ("params pointer probe deferred — "
                     "openfhe.CryptoContext exposes per-attribute "
                     "getters instead of GetCryptoParameters()"),
        }

        # ---- slots ----
        try:
            pt_slots_default = cc.MakeCKKSPackedPlaintext(values)
            pt_slots_perturbed = cc.MakeCKKSPackedPlaintext(values, slots=4)
            slots_default = pt_slots_default.GetSlots()
            slots_perturbed = pt_slots_perturbed.GetSlots()
            slots_block = {
                "default": slots_default,
                "perturbed": slots_perturbed,
                "direction": "changes",
                "matches_prediction": slots_default != slots_perturbed,
            }
        except Exception as exc:
            slots_block = {"error": str(exc), "direction": "changes"}

        result = {
            "method": "MakeCKKSPackedPlaintext",
            "stack": "python",
            "mode": "live_at_9100_0",
            "perturbations": {
                "noiseScaleDeg": noise_block,
                "level": level_block,
                "params": params_block,
                "slots": slots_block,
            },
        }
        json.dump(result, sys.stdout)
        return 0
    except Exception:
        json.dump({
            "method": "MakeCKKSPackedPlaintext",
            "error": "unexpected exception",
            "traceback": traceback.format_exc(),
            "perturbations": {},
        }, sys.stdout)
        return 1


if __name__ == "__main__":
    sys.exit(main())
